ledger/ledger-on-memory/src/main/scala/com/daml/ledger/on/memory/InMemoryLedgerReaderWriter.scala

```diff
@@ -53,40 +49,57 @@ final class InMemoryLedgerReaderWriter(
     extends LedgerWriter
     with LedgerReader {
 
-  private val engine = Engine()
-
   private val currentState = new InMemoryState()
 
-  override def commit(correlationId: String, envelope: Array[Byte]): Future[SubmissionResult] =
-    Future {
-      val submission = Envelope
-        .openSubmission(envelope)
-        .getOrElse(throw new IllegalArgumentException("Not a valid submission in envelope"))
-      currentState.synchronized {
-        val stateInputs: Map[DamlStateKey, Option[DamlStateValue]] =
-          submission.getInputDamlStateList.asScala
-            .map(key => key -> currentState.state.get(key.toByteString))(breakOut)
-        val entryId = sequentialLogEntryId.next()
-        val (logEntry, damlStateUpdates) =
-          KeyValueCommitting.processSubmission(
-            engine,
-            entryId,
-            currentRecordTime(),
-            LedgerReader.DefaultConfiguration,
-            submission,
-            participantId,
-            stateInputs,
-          )
-        val stateUpdates = damlStateUpdates.map {
-          case (damlStateKey, value) => damlStateKey.toByteString -> value
+  private class InMemoryLedgerStateAccess(theParticipantId: ParticipantId)
+      extends LedgerStateAccess {
+    override def inTransaction[T](body: LedgerStateOperations => Future[T]): Future[T] =
+      Future {
+        currentState.state.synchronized {
+          body(new InMemoryLedgerStateOperations)
         }
-        currentState.log += LogEntry(entryId, Envelope.enclose(logEntry).toByteArray)
-        currentState.state ++= stateUpdates
-        dispatcher.signalNewHead(currentState.log.size)
-      }
-      SubmissionResult.Acknowledged
+      }.flatten
+
+    override def participantId: String = theParticipantId
+  }
+
+  private class InMemoryLedgerStateOperations extends LedgerStateOperations {
+    override def readState(key: Array[Byte]): Future[Option[Array[Byte]]] = Future.successful {
+      currentState.state.get(ByteString.copyFrom(key))
     }
 
+    override def writeState(keyValuePairs: Seq[(Array[Byte], Array[Byte])]): Future[Unit] =
+      Future.successful {
+        currentState.state ++= keyValuePairs.map {
+          case (keyBytes, valueBytes) => ByteString.copyFrom(keyBytes) -> valueBytes
+        }
+      }
+
+    override def appendToLog(key: Array[Byte], value: Array[Byte]): Future[Unit] =
+      Future.successful {
+        val damlLogEntryId = KeyValueCommitting.unpackDamlLogEntryId(key)
+        val logEntry = LogEntry(damlLogEntryId, value)
+        val newHead = currentState.log.synchronized {
+          currentState.log += logEntry
+          currentState.log.size
+        }
+        dispatcher.signalNewHead(newHead)
+      }
+  }
+
+  override def commit(correlationId: String, envelope: Array[Byte]): Future[SubmissionResult] = {
+    val validator = SubmissionValidator.create(
```

Comment:
Is there a good reason to recreate the validator for each request separately? Related to my later comment about one engine per validator instance.

Reply:
Not anymore -- changed to reuse the same validator instance for each commit.